redash/query_runner/hive_ds.py

```diff
@@ -97,7 +114,42 @@ def run_query(self, query, user):
 
         connection = None
         try:
-            connection = hive.connect(**self.configuration.to_dict())
+            if self.configuration.get('use_http', False):
+                # default to https
+                scheme = self.configuration.get('http_scheme', 'https')
+
+                # if path is set but is missing initial slash, append it
+                path = self.configuration.get('http_path', '')
+                if path and path[0] != '/':
+                    path = '/' + path
+
+                # if port is set prepend colon
+                port = self.configuration.get('port', '')
+                if port:
+                    port = ':' + port
+
+                # build uri
+                http_uri = "{}://{}{}{}".format(scheme, host, port, path)
+
+                # create transport
+                transport = THttpClient.THttpClient(http_uri)
+
+                # if username or password is set, add Authorization header
+                username = self.configuration.get('username', '')
+                password = self.configuration.get('http_password', '')
+                if username | password:
+                    auth = base64.b64encode(username + ':' + password)
+                    transport.setCustomHeaders({'Authorization': 'Basic ' + auth})
+
+                # create connection
+                connection = hive.connect(thrift_transport=transport, **self.configuration.to_dict())
```

Comment:
Passing the configuration values as keyword arguments won't work, since the [`Connection` class](https://github.com/dropbox/PyHive/blob/d19cb0c2e14fce1888e327755526e94c30cb407d/pyhive/hive.py#L97-L102) which the `hive.connect` function calls, doesn't support values such as `use_http`, `http_scheme`, etc.

So similar to the case in which `use_http` is `False` we'll need to build the parameters to pass to hive.connect individually. Since the actual difference between both calls is the existence of `thrift_transport` in the case of using `use_http` being `True`, let's simply build the parameters once and only update them with the `thrift_transport` value when needed.

Reply:
So there's a couple of things going on here.

1. Issue: @robotblake wrote [this code](https://gist.github.com/robotblake/d0eff602664ab62959772dbb4f0cca22) 6 months ago. 

Action: I went looking for the version of the PyHive code that Blake would have been looking at. I didn't find `thrift_transport` [supported](https://github.com/dropbox/PyHive/blob/a064a496001192bccc47bafdef819ed90ea2529f/pyhive/hive.py) until a July commit which makes me think Blake was working off of a different implementation. Because passing configuration variables of host, port, auth, kerberos_service_name, and/or password as this code does in `**self.configuration.to_dict()` will [raise a ValueError](https://github.com/dropbox/PyHive/blob/a064a496001192bccc47bafdef819ed90ea2529f/pyhive/hive.py#L98-L108) in the Connection class.

2. Issue: The `host` variable used in [line 132](https://github.com/getredash/redash/pull/2747/files#diff-2b97dc03b6ae2c7943fe42487818ade3R132) is, as far as I can tell, never assigned before being used there. 

Action: Add host assignment above that line. 

3. Issue: No support for `use_http` or `http_scheme` in the Connection class.

Action: don't pass `**self.configuration.to_dict()` when `use_http` is True.

4. Issue: Don't rebuild parameters twice.

There's actually only one parameter used in both in the exact same manner and that's `host`, which I've placed above the if statement. Please note the default values between the `use_http` is True case versus the `use_http` is False case are different for port and username. The other configuration values are only used in either one case or the other.

5. Issue: There's a typo in configuration in line 149 below.

Action: Fixed.